source/mir/random/generic/package.d

```diff
@@ -0,0 +1,261 @@
+
+Macros:
+    F_TILDE=g(x)
+*/
+module mir.random.generic;
+
+import mir.random.generic.types : IntervalPoint;
+
+import std.traits : ReturnType;
+import std.random : isUniformRNG;
+
+
+/**
+The Transformed Density Rejection with Inflection Points (Tinflex) algorithm
+can sample from arbitrary distributions given its density function f, its
+first two derivatives and a partitioning into intervals with at most one inflection
+point.
+
+Params:
+    f0 = probability density function of the distribution
+    f1 = first derivative of f0
+    f1 = second derivative of f0
+    c = T_c family
+    points = non-overlapping partitioning with at most one inflection point per interval
+    rho = efficiency of the Tinflex algorithm
+
+Returns:
+    Tinflex Generator.
+*/
+Tinflex!(F0, S) tinflex(F0, F1, F2, S)
+               (in F0 f0, in F1 f1, in F2 f2,
+                S c, S[] points, S rho = 1.1)
+{
+    import mir.random.generic.calc : calcPoints;
+    // pre-calculate all the points
+    auto ips = calcPoints(f0, f1, f2, c, points, 1.1);
+    return Tinflex!(F0, S)(f0, ips, c);
+}
+
+/**
+Data body of the Tinflex algorithm.
+Can be used to sample from the distribution.
+*/
+struct Tinflex(F0, S)
+{
+    // saved internal state of Tinflex:
+
+    // density function
+    private F0 _f0;
+
+    /// density function of the distribution
+    S f0(S x)
+    {
+        return _f0(x);
+    }
+
+    // generated partition points
+    private IntervalPoint!S[] ips;
+
+    // global T_c family
+    private S c;
+
+    private this(F0 f0, IntervalPoint!S[] ips, S c)
+    {
+        this._f0 = f0;
+        this.ips = ips;
+        this.c = c;
+    }
+
+    /**
+    Sample n times from the distribution.
+    Params:
+        n = number of times to samples
+        rng = random number generator to use
+    Returns:
+        Array of length n with the samples
+    */
+    S opCall() const
+    {
+        import std.random : rndGen;
+        return tfSample(_f0, ips, c, rndGen);
+    }
+
+    /// ditto
+    S opCall(RNG)(ref RNG rng) const
+        if (isUniformRNG!RNG)
+    {
+        return tfSample(_f0, ips, c, rng);
+    }
+}
+
+/**
+Sample from the distribution.
+Params:
+    ips = calculated inflection points
+    rng = random number generator to use
+*/
+protected S tfSample(F0, S, RNG)
+          (in F0 f0, in IntervalPoint!S[] ips, in S c, ref RNG rng)
+    if (is(ReturnType!F0 == S) &&
+        isUniformRNG!RNG)
+{
+    import std.algorithm: filter, joiner, map, sum;
+
+    auto s = ips.filter!`a.right != 0`.map!`a.hatA`.sum;
+    auto areas = ips.filter!`a.right != 0`.map!((x) => x.hatA / s);
+
+    import std.random: dice, uniform;
+    import std.math: abs;
+    import mir.internal.math: exp;
+
+    import mir.random.generic.transformations : inverse, antiderivative, inverseAntiderivative;
+
+    double X;
+    // acceptance-rejection sampling
+    while (true)
+    {
+        auto j = dice(rng, areas);
+        auto u = uniform(0, 1, rng);
```

Comment:
BTW I would be a bit suspicious about the quality of the implementation of `dice`.  It's a while since I looked in-depth, but I suspect it may be slow.  It might be better to find another way of doing that.